packages/turbo-r2-archive/src/routes.ts

```diff
@@ -0,0 +1,122 @@
+import { zValidator } from "@hono/zod-validator";
+import { Hono } from "hono";
+import { bearerAuth } from "hono/bearer-auth";
+import { cors } from "hono/cors";
+import { HTTPException } from "hono/http-exception";
+import { z } from "zod";
+import { bustOldCache } from "./autoCacheBust";
+
+export const router = new Hono<{ Bindings: Env }>();
+
+const paramValidator = z.object({
+	artifactID: z.string(),
+	teamID: z.string().optional(),
+});
+const queryValidator = z.object({
+	teamId: z.string().optional(),
+	slug: z.string().optional(),
+});
+
+router.onError((error, c) => {
+	if (error instanceof HTTPException) {
+		return error.getResponse();
+	}
+	return c.json({ ...error }, 500);
+});
+
+router.use("*", cors());
+
+router.use("*", async (c, next) => {
+	const middleware = bearerAuth({ token: c.env.TURBO_TOKEN });
+	await middleware(c, next);
+});
+
+router.post("/artifacts/manual-cache-bust", async (c) => {
+	/**
+	 * manual cache busting, it will bust the entire cache.
+	 */
+	await bustOldCache({
+		...c.env,
+	});
+
+	// maybe this could return the keys that were busted?
+	return c.json({ success: true });
+});
+
+router.put(
+	"/v8/artifacts/:artifactID",
+	zValidator("param", paramValidator),
+	zValidator("query", queryValidator),
+	async (c) => {
+		const { artifactID } = c.req.valid("param");
+		const { teamId, slug } = c.req.valid("query");
+		const teamID = teamId ?? slug;
+
+		if (!teamID) {
+			return c.json({ error: "MISSING_TEAM_ID" }, 400);
+		}
+
+		if (c.req.headers.get("Content-Type") !== "application/octet-stream") {
+			return c.json({ error: "EXPECTED_OCTET_STREAM" }, 415);
+		}
+
+		const artifactTag = c.req.headers.get("x-artifact-tag");
+		const r2Object = await c.env.R2_ARTIFACT_ARCHIVE.put(
+			`${teamID}/${artifactID}`,
+			c.req.body,
+			{
+				customMetadata: artifactTag
+					? {
+							artifactTag,
+					  }
+					: undefined,
+			}
+		);
+
+		return c.json(
+			{ teamID, artifactID, storagePath: r2Object.key, size: r2Object.size },
+			201
+		);
+	}
+);
+
+router.get(
+	"/v8/artifacts/:artifactID",
+	zValidator("param", paramValidator),
+	zValidator("query", queryValidator),
+	async (c) => {
+		const { artifactID } = c.req.valid("param");
+		const { teamId, slug } = c.req.valid("query");
+		const teamID = teamId ?? slug;
+
+		if (!teamID) {
+			return c.json({ error: "MISSING_TEAM_ID & QUERY" }, 400);
+		}
+
+		if (artifactID === "list") {
+			const list = await c.env.R2_ARTIFACT_ARCHIVE.list();
+			return c.json(list.objects);
+		}
+
+		const r2Object = await c.env.R2_ARTIFACT_ARCHIVE.get(
+			`${teamID}/${artifactID}`
+		);
+		if (!r2Object) {
+			return c.notFound();
+		}
+
+		c.header("Content-Type", "application/octet-stream");
+		if (r2Object.customMetadata?.artifactTag) {
```

Comment:
Could this be stored in httpMetadata? That's probably a better place for headers than customMetadata

Reply:
I suppose, what is the difference?

Reply:
Not a big deal, but it gives you this nice API for writing headers
```
const headers = new Headers();
object.writeHttpMetadata(headers);
```
However, it looks like it might only support a [limited subset of headers](https://developers.cloudflare.com/r2/api/workers/workers-api-reference/#http-metadata), so never mind